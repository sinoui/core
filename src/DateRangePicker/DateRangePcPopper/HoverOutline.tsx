import styled from 'styled-components';

import formatDate from '../../DatePicker/formatDate';
import getDayOfWeek from '../helpers/getDayOfWeek';
import getWeekNoOfMonth from '../helpers/getWeekNoOfMonth';
import isSameMonth from '../helpers/isSameMonth';

const BORDER_COLOR: Record<string, string> = {
  light: '#9a999a',
};

const HoverOutlineWrapper = styled.div`
  position: absolute;
  top: 0px;
  left: 0px;
  width: 28px;
  height: 28px;
  box-sizing: border-box;
  border: 1px solid
    ${({ theme }) =>
      BORDER_COLOR[theme.palette.type] ?? theme.palette.text.disabled};
  border-radius: 14px;
  transition: ${({ theme }) =>
    theme.transitions.create('transform', {
      duration: theme.transitions.duration.shortest,
    })};
`;

/**
 * 悬停轮廓组件属性
 */
export interface HoverOutlineProps {
  /**
   * 悬停日期
   */
  hoverDate: Date;
  /**
   * 年份
   */
  year: number;
  /**
   * 月份
   */
  month: number;
  /**
   * 星期开始位置。`0`表示开始的是星期日，`1`表示星期一。默认为`1`。
   */
  startOfWeek?: 0 | 1;
}

/**
 * 悬停轮廓
 *
 * @param props 组件属性
 * @param props.hoverDate 悬停日期
 * @param props.year 年份
 * @param props.month 月份
 * @param props.startOfWeek 星期开始位置。`0`表示开始的是星期日，`1`表示星期一。默认为`1`。
 */
const HoverOutline: React.FC<HoverOutlineProps> = ({
  hoverDate,
  year,
  month,
  startOfWeek,
  ...rest
}) => {
  const weekNo = getWeekNoOfMonth(hoverDate, startOfWeek === 0);
  const dayOfWeek = getDayOfWeek(hoverDate, startOfWeek === 0);

  const x =
    (isSameMonth(hoverDate, year, month) ? 16 : 256 + 16) + dayOfWeek * 32 + 2;
  const y = weekNo * 32 + 2;

  return (
    <HoverOutlineWrapper
      className="sinoui-date-range-picker__hover-outline"
      style={{
        transform: `translate(${x}px, ${y}px)`,
      }}
      data-hover-date={formatDate(hoverDate)}
      {...rest}
    />
  );
};

export default HoverOutline;
