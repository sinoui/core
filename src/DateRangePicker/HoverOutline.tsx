import React from 'react';
import styled from 'styled-components';
import getWeekNoOfMonth from './helpers/getWeekNoOfMonth';
import isSameMonth from './helpers/isSameMonth';
import getDayOfWeek from './helpers/getDayOfWeek';

const HoverOutlineWrapper = styled.div`
  position: absolute;
  top: 0px;
  left: 0px;
  width: 28px;
  height: 28px;
  box-sizing: border-box;
  border: 1px solid ${({ theme }) => theme.palette.text.disabled};
  border-radius: 14px;
  transition: ${({ theme }) =>
    theme.transitions.create('transform', {
      duration: theme.transitions.duration.shortest,
    })};
`;

/**
 * 悬停轮廓
 */
const HoverOutline = ({
  hoverDate,
  year,
  month,
  ...rest
}: {
  hoverDate: Date;
  year: number;
  month: number;
}) => {
  const weekNo = getWeekNoOfMonth(hoverDate);
  const dayOfWeek = getDayOfWeek(hoverDate);

  const x =
    (isSameMonth(hoverDate, year, month) ? 16 : 256 + 16) + dayOfWeek * 32 + 2;
  const y = weekNo * 32 + 2;

  return (
    <HoverOutlineWrapper
      className="sinoui-date-range-picker__hover-outline"
      style={{
        transform: `translate(${x}px, ${y}px)`,
      }}
      {...rest}
    />
  );
};

export default HoverOutline;
