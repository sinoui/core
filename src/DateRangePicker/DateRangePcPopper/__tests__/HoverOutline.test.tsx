import React from 'react';
import { cleanup, render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import renderer from 'react-test-renderer';
import HoverOutline from '../HoverOutline';

afterEach(cleanup);

it('展现悬停轮廓', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <HoverOutline
        data-testid="hover-outline"
        hoverDate={new Date(2020, 6, 8)}
        year={2020}
        month={6}
      />
    </ThemeProvider>,
  );

  expect(getByTestId('hover-outline')).toHaveClass(
    'sinoui-date-range-picker__hover-outline',
  );
  expect(getByTestId('hover-outline')).toHaveStyle(
    'transform: translate(82px, 34px)',
  );
});

it('悬停轮廓在第二个日历上', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <HoverOutline
        data-testid="hover-outline"
        hoverDate={new Date(2020, 6, 8)}
        year={2020}
        month={7}
      />
    </ThemeProvider>,
  );

  expect(getByTestId('hover-outline')).toHaveStyle(
    'transform: translate(338px, 34px)',
  );
});

it('星期从周日开始', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <HoverOutline
        data-testid="hover-outline"
        hoverDate={new Date(2020, 6, 8)}
        year={2020}
        month={6}
        startOfWeek={0}
      />
    </ThemeProvider>,
  );

  expect(getByTestId('hover-outline')).toHaveStyle(
    'transform: translate(114px, 34px)',
  );
});

it('快照测试', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <HoverOutline
          data-testid="hover-outline"
          hoverDate={new Date(2020, 6, 8)}
          year={2020}
          month={6}
        />
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
