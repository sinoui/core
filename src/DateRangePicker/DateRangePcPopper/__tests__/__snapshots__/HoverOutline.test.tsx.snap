// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 1`] = `
<div
  className="sc-bdnylx kLmPJw sinoui-date-range-picker__hover-outline"
  data-hover-date="2020-07-08"
  data-testid="hover-outline"
  style={
    Object {
      "transform": "translate(82px, 34px)",
    }
  }
/>
`;
